File: test-suite-python/src/test/python/helloworld/HealthCheckTest.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Bean, Factory
from micronaut.grpc.annotation import GrpcChannel
from micronaut.grpc.server import GrpcServerChannel
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .HealthService import HealthService

try:
    from io.grpc import ManagedChannel
    from io.grpc.health.v1 import HealthCheckRequest, HealthCheckResponse, HealthGrpc
    from io.grpc.protobuf.services import HealthStatusManager
except ImportError:  # TODO(python): packages under `io.` other than `io.micronaut` cannot be imported at runtime
    from grpc import ManagedChannel
    from grpc.health.v1 import HealthCheckRequest, HealthCheckResponse, HealthGrpc
    from grpc.protobuf.services import HealthStatusManager


@MicronautTest
class HealthCheckTest:

    health_stub: Annotated[HealthGrpc.HealthBlockingStub, Inject]
    health_service: Annotated[HealthService, Inject]

    @Test
    def test_health(self):
        assert self.health_stub.check(HealthCheckRequest.newBuilder().build()).getStatus() == HealthCheckResponse.ServingStatus.SERVING

        self.health_service.set_status(HealthStatusManager.SERVICE_NAME_ALL_SERVICES, HealthCheckResponse.ServingStatus.NOT_SERVING)

        assert self.health_stub.check(HealthCheckRequest.newBuilder().build()).getStatus() == HealthCheckResponse.ServingStatus.NOT_SERVING


@Factory
class HealthClients:

    @Bean
    def health_stub(self, channel: Annotated[ManagedChannel, GrpcChannel(GrpcServerChannel.NAME)]) -> HealthGrpc.HealthBlockingStub:
        return HealthGrpc.newBlockingStub(channel)

File: test-suite/src/test/java/helloworld/HealthCheckTest.java
package helloworld;

import io.grpc.ManagedChannel;
import io.grpc.health.v1.HealthCheckRequest;
import io.grpc.health.v1.HealthCheckResponse;
import io.grpc.health.v1.HealthGrpc;
import io.grpc.protobuf.services.HealthStatusManager;
import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.grpc.annotation.GrpcChannel;
import io.micronaut.grpc.server.GrpcServerChannel;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;

import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;

import static org.junit.jupiter.api.Assertions.assertEquals;

@MicronautTest
class HealthCheckTest {

    @Inject
    HealthGrpc.HealthBlockingStub healthStub;

    @Inject
    HealthService healthService;

    @Test
    void testHealth() {
        assertEquals(
            HealthCheckResponse.ServingStatus.SERVING,
            healthStub.check(HealthCheckRequest.newBuilder().build()).getStatus()
        );

        healthService.setStatus(HealthStatusManager.SERVICE_NAME_ALL_SERVICES, HealthCheckResponse.ServingStatus.NOT_SERVING);

        assertEquals(
            HealthCheckResponse.ServingStatus.NOT_SERVING,
            healthStub.check(HealthCheckRequest.newBuilder().build()).getStatus()
        );
    }

    @Factory
    static class HealthClients {

        @Bean
        HealthGrpc.HealthBlockingStub blockingStub(@GrpcChannel(GrpcServerChannel.NAME) ManagedChannel channel) {
            return HealthGrpc.newBlockingStub(channel);
        }
    }
}
